### PR TITLE
chore: create ResourceNotificationMessage and AsNotifier

### LINKED ISSUE #15213: chore(rbac): add dbauthz.AsNotifier
Context: https://github.com/coder/coder/pull/15195#discussion_r1814644800

We should add a separate RBAC role specifically for enqueuing notifications instead of using `dbauthz.AsSystemRestricted`

### PR DESCRIPTION
**Pull request by DanielleMaywood**
Closes https://github.com/coder/coder/issues/15213

This PR enables sending notifications without requiring the auth system context, instead using a new auth notifier context.

### CODE REVIEW

**Review by mafredri — APPROVED** [review]
LGTM but I'll defer to @johnstcn or @Emyrk for final approval for all things RBAC 😄

File: cli/server.go [hunk 1]
```diff
@@ -910,8 +910,8 @@ func (r *RootCmd) Server(newAPI func(context.Context, *coderd.Options) (*coderd.
 				return xerrors.Errorf("failed to instantiate notification manager: %w", err)
 			}
 
-			// nolint:gocritic // TODO: create own role.
-			notificationsManager.Run(dbauthz.AsSystemRestricted(ctx))
+			// nolint:gocritic // We need to run the manager in a notifier context.
```
Comment (dannykopping) [inline cli/server.go:913]: Does this lint error occur when not using system?

File: coderd/database/dbauthz/dbauthz.go [hunk 2]
```diff
@@ -327,6 +345,11 @@ func AsKeyReader(ctx context.Context) context.Context {
 	return context.WithValue(ctx, authContextKey{}, subjectCryptoKeyReader)
 }
 
+// AsNotifier returns a context with an actor that has permissions required for creating notifications.
```
Comment (dannykopping) [inline coderd/database/dbauthz/dbauthz.go:348]: This comment is inaccurate; it's not just for creating but fully CRUDing

**Review by dannykopping — APPROVED** [review]
Thanks very much for adding this 👍

File: cli/server.go (hunk 1 above)
Comment (DanielleMaywood) [reply cli/server.go:913]: Yep, it occurs for any of the `authz.As{Role}` functions

https://github.com/coder/coder/blob/6e54bd918387c7b591f47b82dcbd7317f39e4e95/scripts/rules.go#L23-L44

Comment (dannykopping) [reply cli/server.go:913]: Ha, TIL.

**Pull request merged**